File: policy-engine/sdk/dotnet/src/AgentControlSpecification/Telemetry.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System.Diagnostics;
using System.Diagnostics.Metrics;
using System.Text;
using System.Text.Json;

namespace AgentControlSpecification;

public enum TelemetryEventType
{
    Decision,
    AnnotatorDispatch,
    PolicyEvaluation,
    EvaluationTiming,
    InterventionPointTransformed,
    AnnotatorFailed,
    PolicyFailed,
}

public static class TelemetryEventTypeExtensions
{
    public static string ToWireName(this TelemetryEventType eventType) => eventType switch
    {
        TelemetryEventType.Decision => "decision",
        TelemetryEventType.AnnotatorDispatch => "annotator_dispatch",
        TelemetryEventType.PolicyEvaluation => "policy_evaluation",
        TelemetryEventType.EvaluationTiming => "evaluation_timing",
        TelemetryEventType.InterventionPointTransformed => "intervention_point.transformed",
        TelemetryEventType.AnnotatorFailed => "annotator_failed",
        TelemetryEventType.PolicyFailed => "policy_failed",
        _ => throw new ArgumentOutOfRangeException(nameof(eventType), eventType, "Unknown Agent Control Specification telemetry event type."),
    };
}

public static class TelemetryRedaction
{
    private const int MaxReasonCodeUtf8Bytes = 96;

    public static string? SafeReasonCode(string? reason)
    {
        if (reason is null)
        {
            return null;
        }

        return IsIdentifierReasonCode(reason) ? reason : "policy_reason";
    }

    public static string? ErrorClassFor(string? reason) =>
        reason is not null && reason.StartsWith("runtime_error:", StringComparison.Ordinal)
            ? "runtime_error"
            : null;

    private static bool IsIdentifierReasonCode(string reason)
    {
        if (reason.Length == 0 || Encoding.UTF8.GetByteCount(reason) > MaxReasonCodeUtf8Bytes)
        {
            return false;
        }

        foreach (var ch in reason)
        {
            if (ch > 0x7f)
            {
                return false;
            }

            var isAllowed =
                (ch >= 'A' && ch <= 'Z') ||
                (ch >= 'a' && ch <= 'z') ||
                (ch >= '0' && ch <= '9') ||
                ch is '_' or '-' or '.' or ':' or '/';
            if (!isAllowed)
            {
                return false;
            }
        }

        return true;
    }
}

public sealed record TelemetryEvent
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    public TelemetryEvent(
        TelemetryEventType eventType,
        InterventionPoint interventionPoint,
        Decision? decision = null,
        string? reasonCode = null,
        string? errorClass = null,
        string? policyId = null,
        IReadOnlyList<string>? annotators = null,
        EnforcementMode? enforcementMode = null,
        double? durationMs = null,
        string? evidenceArtefact = null,
        IReadOnlyList<string>? evidenceVerificationPointerKeys = null,
        string? actionIdentity = null,
        IReadOnlyDictionary<string, string>? metadata = null)
    {
        EventType = eventType;
        InterventionPoint = interventionPoint;
        Decision = decision;
        ReasonCode = reasonCode;
        ErrorClass = errorClass;
        PolicyId = policyId;
        Annotators = CopyList(annotators);
        EnforcementMode = enforcementMode;
        DurationMs = durationMs;
        EvidenceArtefact = evidenceArtefact;
        EvidenceVerificationPointerKeys = CopyList(evidenceVerificationPointerKeys);
        ActionIdentity = actionIdentity;
        Metadata = metadata is null
            ? new Dictionary<string, string>(StringComparer.Ordinal)
            : new Dictionary<string, string>(metadata, StringComparer.Ordinal);
    }

    public TelemetryEventType EventType { get; init; }

    public InterventionPoint InterventionPoint { get; init; }

    public Decision? Decision { get; init; }

    public string? ReasonCode { get; init; }

    public string? ErrorClass { get; init; }

    public string? PolicyId { get; init; }

    public IReadOnlyList<string> Annotators { get; init; }

    public EnforcementMode? EnforcementMode { get; init; }

    public double? DurationMs { get; init; }

    public string? EvidenceArtefact { get; init; }

    public IReadOnlyList<string> EvidenceVerificationPointerKeys { get; init; }

    public string? ActionIdentity { get; init; }

    public IReadOnlyDictionary<string, string> Metadata { get; init; }

    public static TelemetryEvent FromResult(
        InterventionPoint interventionPoint,
        EnforcementMode? mode,
        InterventionPointResult result,
        double? durationMs,
        string? policyId = null,
        IReadOnlyList<string>? annotators = null)
    {
        var verdict = result.Verdict;
        var evidence = verdict.Evidence;
        var pointerKeys = evidence?.VerificationPointers is null
            ? Array.Empty<string>()
            : evidence.VerificationPointers.Keys.OrderBy(key => key, StringComparer.Ordinal).ToArray();
        var resolvedAnnotators = annotators ?? AnnotatorNamesFromResult(result.PolicyInput);

        return new TelemetryEvent(
            TelemetryEventType.Decision,
            interventionPoint,
            verdict.Decision,
            TelemetryRedaction.SafeReasonCode(verdict.Reason),
            TelemetryRedaction.ErrorClassFor(verdict.Reason),
            policyId,
            resolvedAnnotators,
            mode,
            durationMs,
            evidence?.Artefact,
            pointerKeys,
            result.ActionIdentity,
            new Dictionary<string, string>(StringComparer.Ordinal));
    }

    // Sorted annotator names from the result policy input, used when the caller
    // has no manifest-configured annotator list (for example a from_url or YAML
    // manifest source). Mirrors the Python and Node host layers; reads the
    // annotation KEYS only, never the annotator output values.
    private static IReadOnlyList<string> AnnotatorNamesFromResult(JsonElement? policyInput)
    {
        if (policyInput is not JsonElement element || element.ValueKind != JsonValueKind.Object)
        {
            return Array.Empty<string>();
        }

        if (!element.TryGetProperty("annotations", out var annotations) ||
            annotations.ValueKind != JsonValueKind.Object)
        {
            return Array.Empty<string>();
        }

        return annotations
            .EnumerateObject()
            .Select(property => property.Name)
            .OrderBy(name => name, StringComparer.Ordinal)
            .ToArray();
    }

    public IReadOnlyDictionary<string, object?> ToDictionary() => new Dictionary<string, object?>(StringComparer.Ordinal)
    {
        ["event_type"] = EventType.ToWireName(),
        ["intervention_point"] = InterventionPoint.ToWireName(),
        ["decision"] = Decision?.ToWireName(),
        ["reason_code"] = ReasonCode,
        ["error_class"] = ErrorClass,
        ["policy_id"] = PolicyId,
        ["annotators"] = (Annotators ?? Array.Empty<string>()).ToArray(),
        ["enforcement_mode"] = EnforcementMode?.ToWireName(),
        ["duration_ms"] = DurationMs,
        ["evidence_artefact"] = EvidenceArtefact,
        ["evidence_verification_pointer_keys"] = (EvidenceVerificationPointerKeys ?? Array.Empty<string>()).ToArray(),
        ["action_identity"] = ActionIdentity,
        ["metadata"] = Metadata is null
            ? new Dictionary<string, string>(StringComparer.Ordinal)
            : new Dictionary<string, string>(Metadata, StringComparer.Ordinal),
    };

    public string ToJsonString() => JsonSerializer.Serialize(ToDictionary(), JsonOptions);

    private static IReadOnlyList<string> CopyList(IReadOnlyList<string>? values) =>
        values is null ? Array.Empty<string>() : values.ToArray();
}

public interface ITelemetrySink
{
    void Emit(TelemetryEvent telemetryEvent);

    void ForceFlush();

    void Shutdown();
}

/// <summary>
/// Per intervention point telemetry labels resolved from the fully merged
/// manifest by the native core. Keys are intervention point wire names.
/// </summary>
public sealed record PolicyLabelMap(
    IReadOnlyDictionary<string, string> PolicyIds,
    IReadOnlyDictionary<string, IReadOnlyList<string>> Annotators)
{
    public static PolicyLabelMap Empty { get; } = new(
        new Dictionary<string, string>(StringComparer.Ordinal),
        new Dictionary<string, IReadOnlyList<string>>(StringComparer.Ordinal));

    /// <summary>
    /// Parse the native <c>acs_runtime_policy_labels</c> JSON payload of shape
    /// <c>{ "&lt;point&gt;": { "policy_id": string|null, "annotators": [string] } }</c>
    /// into the two lookup indexes the telemetry funnel reads.
    /// </summary>
    public static PolicyLabelMap FromJson(string json)
    {
        var policyIds = new Dictionary<string, string>(StringComparer.Ordinal);
        var annotators = new Dictionary<string, IReadOnlyList<string>>(StringComparer.Ordinal);
        using var document = JsonDocument.Parse(json);
        if (document.RootElement.ValueKind != JsonValueKind.Object)
        {
            return Empty;
        }

        foreach (var point in document.RootElement.EnumerateObject())
        {
            if (point.Value.ValueKind != JsonValueKind.Object)
            {
                continue;
            }

            if (point.Value.TryGetProperty("policy_id", out var policyId)
                && policyId.ValueKind == JsonValueKind.String)
            {
                policyIds[point.Name] = policyId.GetString() ?? string.Empty;
            }

            if (point.Value.TryGetProperty("annotators", out var annotatorNames)
                && annotatorNames.ValueKind == JsonValueKind.Array)
            {
                var names = annotatorNames
                    .EnumerateArray()
                    .Where(name => name.ValueKind == JsonValueKind.String)
                    .Select(name => name.GetString() ?? string.Empty)
                    .Where(name => name.Length > 0)
                    .OrderBy(name => name, StringComparer.Ordinal)
                    .ToArray();
                if (names.Length > 0)
                {
                    annotators[point.Name] = names;
                }
            }
        }

        return new PolicyLabelMap(policyIds, annotators);
    }
}

/// <summary>
/// Implemented by a runtime that can expose the merged manifest's telemetry
/// labels. The host telemetry index is populated from this on construction so
/// <c>policy_id</c> and <c>annotators</c> are present for every constructor,
/// including <c>FromManifestChain</c> where the host never parses manifest text.
/// </summary>
public interface IPolicyLabelSource
{
    PolicyLabelMap PolicyLabels();
}

public sealed class InMemoryTelemetrySink : ITelemetrySink
{
    private readonly object gate = new();
    private readonly List<TelemetryEvent> events = [];

    // Returns a snapshot under the lock so concurrent Emit calls (evaluations
    // funnel through Task.Run, so post-await emits can land on different
    // thread-pool threads) cannot tear a read. Mirrors the Rust core
    // InMemoryTelemetrySink, whose events() clones under a Mutex.
    public IReadOnlyList<TelemetryEvent> Events
    {
        get
        {
            lock (gate)
            {
                return events.ToArray();
            }
        }
    }

    public void Emit(TelemetryEvent telemetryEvent)
    {
        lock (gate)
        {
            events.Add(telemetryEvent);
        }
    }

    public void ForceFlush()
    {
    }

    public void Shutdown()
    {
    }

    public void Clear()
    {
        lock (gate)
        {
            events.Clear();
        }
    }
}

public sealed class JsonStdoutTelemetrySink : ITelemetrySink
{
    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);
    private readonly object gate = new();
    private readonly TextWriter writer;

    public JsonStdoutTelemetrySink(TextWriter? writer = null)
    {
        this.writer = writer ?? Console.Out;
    }

    // Serialize outside the lock (no shared state), then guard the write and
    // flush so concurrent emits cannot interleave partial lines on the shared
    // writer. Mirrors the Rust core StdoutJsonTelemetrySink, whose writer is a
    // Mutex.
    public void Emit(TelemetryEvent telemetryEvent)
    {
        var line = JsonSerializer.Serialize(telemetryEvent.ToDictionary(), JsonOptions);
        lock (gate)
        {
            writer.WriteLine(line);
        }
    }

    public void ForceFlush()
    {
        lock (gate)
        {
            writer.Flush();
        }
    }

    public void Shutdown() => ForceFlush();
}

public sealed class MultiSink : ITelemetrySink
{
    private readonly ITelemetrySink[] sinks;

    public MultiSink(params ITelemetrySink[] sinks)
        : this((IEnumerable<ITelemetrySink>)sinks)
    {
    }

    public MultiSink(IEnumerable<ITelemetrySink> sinks)
    {
        ArgumentNullException.ThrowIfNull(sinks);
        this.sinks = sinks.Select(sink => sink ?? throw new ArgumentException("Telemetry sink collection must not contain null.", nameof(sinks))).ToArray();
    }

    public IReadOnlyList<ITelemetrySink> Sinks => sinks;

    public void Emit(TelemetryEvent telemetryEvent)
    {
        foreach (var sink in sinks)
        {
            try
            {
                sink.Emit(telemetryEvent);
            }
            catch (Exception exception) when (!TelemetryExceptions.IsFatal(exception))
            {
                Trace.TraceWarning($"ACS telemetry sink {sink.GetType().Name} raised during emit: {exception}");
            }
        }
    }

    public void ForceFlush()
    {
        foreach (var sink in sinks)
        {
            try
            {
                sink.ForceFlush();
            }
            catch (Exception exception) when (!TelemetryExceptions.IsFatal(exception))
            {
                Trace.TraceWarning($"ACS telemetry sink {sink.GetType().Name} raised during force flush: {exception}");
            }
        }
    }

    public void Shutdown()
    {
        foreach (var sink in sinks)
        {
            try
            {
                sink.Shutdown();
            }
            catch (Exception exception) when (!TelemetryExceptions.IsFatal(exception))
            {
                Trace.TraceWarning($"ACS telemetry sink {sink.GetType().Name} raised during shutdown: {exception}");
            }
        }
    }
}

// Fatal CLR exceptions must propagate even from the best-effort telemetry and
// manifest-indexing paths that are otherwise designed never to be load-bearing.
// Everything else is caught so a sink or a malformed manifest cannot break the
// host. This mirrors the Python `except Exception` (which excludes the
// fatal `BaseException` interrupts) and the Rust catch_unwind isolation.
internal static class TelemetryExceptions
{
    public static bool IsFatal(Exception exception) =>
        exception is OutOfMemoryException or StackOverflowException;
}

public sealed class OtelMetricsTelemetrySink : ITelemetrySink, IDisposable
{
    public const string DefaultOtelMeterName = "agent_control_specification";

    private static readonly string[] DecisionWireStrings = ["allow", "deny", "warn", "escalate", "transform"];
    private readonly Meter meter;
    private readonly Dictionary<string, Counter<double>> decisionCounters = new(StringComparer.Ordinal);
    private readonly Histogram<double> durationHistogram;
    private bool disposed;

    public OtelMetricsTelemetrySink(string meterName = DefaultOtelMeterName)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(meterName);
        MeterName = meterName;
        meter = new Meter(meterName);
        foreach (var decision in DecisionWireStrings)
        {
            decisionCounters[decision] = meter.CreateCounter<double>($"acs_intervention_{decision}_total");
        }

        durationHistogram = meter.CreateHistogram<double>("acs_intervention_duration_ms");
    }

    public string MeterName { get; }

    public void Emit(TelemetryEvent telemetryEvent)
    {
        if (disposed)
        {
            return;
        }

        // Record one increment and one duration sample per evaluation. Only the
        // base decision event records metrics, matching the Rust OTel sink, so a
        // non-decision event fed in directly cannot double-count.
        if (telemetryEvent.EventType != TelemetryEventType.Decision)
        {
            return;
        }

        var tags = MetricTags(telemetryEvent);
        if (telemetryEvent.Decision is not null)
        {
            var decision = telemetryEvent.Decision.Value.ToWireName();
            if (decisionCounters.TryGetValue(decision, out var counter))
            {
                counter.Add(1.0, tags);
            }
        }

        if (telemetryEvent.DurationMs is not null)
        {
            durationHistogram.Record(telemetryEvent.DurationMs.Value, tags);
        }
    }

    public void ForceFlush()
    {
    }

    public void Shutdown() => Dispose();

    public void Dispose()
    {
        if (disposed)
        {
            return;
        }

        meter.Dispose();
        disposed = true;
    }

    private static TagList MetricTags(TelemetryEvent telemetryEvent)
    {
        var tags = new TagList
        {
            { "event_type", telemetryEvent.EventType.ToWireName() },
            { "intervention_point", telemetryEvent.InterventionPoint.ToWireName() },
        };
        AddIfPresent(ref tags, "enforcement_mode", telemetryEvent.EnforcementMode?.ToWireName());
        AddIfPresent(ref tags, "decision", telemetryEvent.Decision?.ToWireName());
        AddIfPresent(ref tags, "reason_code", telemetryEvent.ReasonCode);
        AddIfPresent(ref tags, "error_class", telemetryEvent.ErrorClass);
        AddIfPresent(ref tags, "policy_id", telemetryEvent.PolicyId);
        if (telemetryEvent.Annotators.Count > 0)
        {
            tags.Add("annotators", string.Join(",", telemetryEvent.Annotators));
        }

        AddIfPresent(ref tags, "evidence_artefact", telemetryEvent.EvidenceArtefact);
        if (telemetryEvent.EvidenceVerificationPointerKeys.Count > 0)
        {
            tags.Add("evidence_verification_pointer_keys", string.Join(",", telemetryEvent.EvidenceVerificationPointerKeys));
        }

        return tags;
    }

    private static void AddIfPresent(ref TagList tags, string key, string? value)
    {
        if (value is not null)
        {
            tags.Add(key, value);
        }
    }
}

File: policy-engine/sdk/node/src/telemetry.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import type {
  Decision,
  EnforcementMode,
  InterventionPoint,
  InterventionPointResult,
  JsonValue,
} from "./index";

export const DEFAULT_OTEL_METER_NAME = "agent_control_specification";

const DECISION_WIRE_STRINGS = ["allow", "deny", "warn", "escalate", "transform"] as const;
const MAX_REASON_CODE_BYTES = 96;
const REASON_CODE_EXTRA_CHARS = new Set(["_", "-", ".", ":", "/"]);

export const TelemetryEventType = Object.freeze({
  Decision: "decision",
  AnnotatorDispatch: "annotator_dispatch",
  PolicyEvaluation: "policy_evaluation",
  EvaluationTiming: "evaluation_timing",
  InterventionPointTransformed: "intervention_point.transformed",
  AnnotatorFailed: "annotator_failed",
  PolicyFailed: "policy_failed",
} as const);
export type TelemetryEventType = (typeof TelemetryEventType)[keyof typeof TelemetryEventType];

export interface TelemetryEventFields {
  eventType: TelemetryEventType;
  interventionPoint: InterventionPoint | string;
  decision?: Decision | null;
  reasonCode?: string | null;
  errorClass?: string | null;
  policyId?: string | null;
  annotators?: readonly string[];
  enforcementMode?: EnforcementMode | null;
  durationMs?: number | null;
  evidenceArtefact?: string | null;
  evidenceVerificationPointerKeys?: readonly string[];
  actionIdentity?: string | null;
  metadata?: Record<string, string>;
}

export interface TelemetryEventObject {
  eventType: TelemetryEventType;
  interventionPoint: InterventionPoint | string;
  decision: Decision | null;
  reasonCode: string | null;
  errorClass: string | null;
  policyId: string | null;
  annotators: string[];
  enforcementMode: EnforcementMode | null;
  durationMs: number | null;
  evidenceArtefact: string | null;
  evidenceVerificationPointerKeys: string[];
  actionIdentity: string | null;
  metadata: Record<string, string>;
}

export class TelemetryEvent {
  readonly eventType: TelemetryEventType;
  readonly interventionPoint: InterventionPoint | string;
  readonly decision: Decision | null;
  readonly reasonCode: string | null;
  readonly errorClass: string | null;
  readonly policyId: string | null;
  readonly annotators: readonly string[];
  readonly enforcementMode: EnforcementMode | null;
  readonly durationMs: number | null;
  readonly evidenceArtefact: string | null;
  readonly evidenceVerificationPointerKeys: readonly string[];
  readonly actionIdentity: string | null;
  readonly metadata: Readonly<Record<string, string>>;

  constructor(fields: TelemetryEventFields) {
    this.eventType = fields.eventType;
    this.interventionPoint = fields.interventionPoint;
    this.decision = fields.decision ?? null;
    this.reasonCode = fields.reasonCode ?? null;
    this.errorClass = fields.errorClass ?? null;
    this.policyId = fields.policyId ?? null;
    this.annotators = [...(fields.annotators ?? [])];
    this.enforcementMode = fields.enforcementMode ?? null;
    this.durationMs = fields.durationMs ?? null;
    this.evidenceArtefact = fields.evidenceArtefact ?? null;
    this.evidenceVerificationPointerKeys = [...(fields.evidenceVerificationPointerKeys ?? [])];
    this.actionIdentity = fields.actionIdentity ?? null;
    this.metadata = { ...(fields.metadata ?? {}) };
  }

  static fromResult(
    interventionPoint: InterventionPoint | string,
    mode: EnforcementMode | null,
    result: InterventionPointResult,
    durationMs: number | null,
    policyId: string | null = null,
    annotators?: readonly string[],
  ): TelemetryEvent {
    const verdict = result.verdict;
    const reason = verdict.reason;
    const evidence = verdict.evidence;
    const verificationPointers = evidence?.verificationPointers;
    const pointerKeys = isRecord(verificationPointers)
      ? Object.keys(verificationPointers).sort(compareUnicodeScalarKeys)
      : [];
    const artefact = evidence?.artefact;
    if (artefact !== undefined && artefact !== null && typeof artefact !== "string") {
      throw new TypeError("evidence artefact must be a string when provided");
    }
    return new TelemetryEvent({
      eventType: TelemetryEventType.Decision,
      interventionPoint,
      decision: verdict.decision,
      reasonCode: safeReasonCode(reason),
      errorClass: errorClassFor(reason),
      policyId,
      annotators: annotators ?? annotatorNames(result.policyInput),
      enforcementMode: mode,
      durationMs,
      evidenceArtefact: artefact ?? null,
      evidenceVerificationPointerKeys: pointerKeys,
      actionIdentity: result.actionIdentity ?? null,
      metadata: {},
    });
  }

  toObject(): TelemetryEventObject {
    return {
      eventType: this.eventType,
      interventionPoint: this.interventionPoint,
      decision: this.decision,
      reasonCode: this.reasonCode,
      errorClass: this.errorClass,
      policyId: this.policyId,
      annotators: [...this.annotators],
      enforcementMode: this.enforcementMode,
      durationMs: this.durationMs,
      evidenceArtefact: this.evidenceArtefact,
      evidenceVerificationPointerKeys: [...this.evidenceVerificationPointerKeys],
      actionIdentity: this.actionIdentity,
      metadata: { ...this.metadata },
    };
  }

  // Wire serialization with snake_case keys, matching the Rust, Python, and
  // .NET sinks so a mixed fleet writes one consistent audit.jsonl shape. The
  // camelCase toObject() stays the idiomatic in-memory view. JSON.stringify of
  // a TelemetryEvent uses this method.
  toJSON(): Record<string, JsonValue> {
    return {
      event_type: this.eventType,
      intervention_point: this.interventionPoint,
      decision: this.decision,
      reason_code: this.reasonCode,
      error_class: this.errorClass,
      policy_id: this.policyId,
      annotators: [...this.annotators],
      enforcement_mode: this.enforcementMode,
      duration_ms: this.durationMs,
      evidence_artefact: this.evidenceArtefact,
      evidence_verification_pointer_keys: [...this.evidenceVerificationPointerKeys],
      action_identity: this.actionIdentity,
      metadata: { ...this.metadata },
    };
  }
}

export function safeReasonCode(reason: string | null | undefined): string | null {
  if (reason === null || reason === undefined) {
    return null;
  }
  if (typeof reason !== "string") {
    throw new TypeError("reason must be a string when provided");
  }
  if (isIdentifierReasonCode(reason)) {
    return reason;
  }
  return "policy_reason";
}

export function errorClassFor(reason: string | null | undefined): string | null {
  if (reason !== null && reason !== undefined && typeof reason !== "string") {
    throw new TypeError("reason must be a string when provided");
  }
  return reason?.startsWith("runtime_error:") ? "runtime_error" : null;
}

function isIdentifierReasonCode(reason: string): boolean {
  if (reason.length === 0 || Buffer.byteLength(reason, "utf8") > MAX_REASON_CODE_BYTES) {
    return false;
  }
  for (const char of reason) {
    const code = char.charCodeAt(0);
    if (code >= 128) {
      return false;
    }
    const isAsciiAlphaNumeric =
      (code >= 48 && code <= 57) ||
      (code >= 65 && code <= 90) ||
      (code >= 97 && code <= 122);
    if (!isAsciiAlphaNumeric && !REASON_CODE_EXTRA_CHARS.has(char)) {
      return false;
    }
  }
  return true;
}

export interface TelemetrySink {
  emit(event: TelemetryEvent): void;
  forceFlush(): void;
  shutdown(): void;
}

export class InMemoryTelemetrySink implements TelemetrySink {
  readonly events: TelemetryEvent[] = [];

  emit(event: TelemetryEvent): void {
    this.events.push(event);
  }

  forceFlush(): void {}

  shutdown(): void {}

  clear(): void {
    this.events.length = 0;
  }
}

export class JsonStdoutTelemetrySink implements TelemetrySink {
  constructor(private readonly stream: NodeJS.WritableStream = process.stdout) {}

  emit(event: TelemetryEvent): void {
    this.stream.write(`${JSON.stringify(event.toJSON())}\n`);
  }

  forceFlush(): void {
    const maybeFlush = (this.stream as { flush?: () => void }).flush;
    if (typeof maybeFlush === "function") {
      maybeFlush.call(this.stream);
    }
  }

  shutdown(): void {
    this.forceFlush();
  }
}

export class MultiSink implements TelemetrySink {
  readonly sinks: readonly TelemetrySink[];

  constructor(sinks: readonly TelemetrySink[]) {
    this.sinks = [...sinks];
  }

  emit(event: TelemetryEvent): void {
    for (const sink of this.sinks) {
      try {
        sink.emit(event);
      } catch (error) {
        console.warn(`Telemetry sink ${sink.constructor?.name ?? "unknown"} raised in emit.`, error);
      }
    }
  }

  forceFlush(): void {
    for (const sink of this.sinks) {
      try {
        sink.forceFlush?.();
      } catch (error) {
        console.warn(`Telemetry sink ${sink.constructor?.name ?? "unknown"} raised in forceFlush.`, error);
      }
    }
  }

  shutdown(): void {
    for (const sink of this.sinks) {
      try {
        sink.shutdown?.();
      } catch (error) {
        console.warn(`Telemetry sink ${sink.constructor?.name ?? "unknown"} raised in shutdown.`, error);
      }
    }
  }
}

type OtelCounter = {
  add(value: number, attributes?: Record<string, string>): void;
};

type OtelHistogram = {
  record(value: number, attributes?: Record<string, string>): void;
};

type OtelMeter = {
  createCounter(name: string): OtelCounter;
  createHistogram(name: string): OtelHistogram;
};

type OtelMeterProvider = {
  getMeter(name: string): OtelMeter;
  forceFlush?: () => void;
  shutdown?: () => void;
};

export interface OtelMetricsTelemetrySinkOptions {
  meterProvider?: OtelMeterProvider;
}

export class OtelMetricsTelemetrySink implements TelemetrySink {
  private static importWarningEmitted = false;
  private readonly meterProvider: OtelMeterProvider | undefined;
  private readonly resolvedMeterProvider: OtelMeterProvider | undefined;
  private readonly decisionCounters = new Map<string, OtelCounter>();
  private durationHistogram: OtelHistogram | undefined;
  private isAvailable = false;
  readonly meterName: string;

  constructor(
    meterName: string = DEFAULT_OTEL_METER_NAME,
    options: OtelMetricsTelemetrySinkOptions = {},
  ) {
    this.meterName = meterName;
    this.meterProvider = options.meterProvider;
    let meter: OtelMeter;
    try {
      const otelApi = require("@opentelemetry/api") as {
        metrics: { getMeter(name: string): OtelMeter; getMeterProvider(): OtelMeterProvider };
      };
      this.resolvedMeterProvider = this.meterProvider ?? otelApi.metrics.getMeterProvider();
      meter = this.meterProvider?.getMeter(meterName) ?? otelApi.metrics.getMeter(meterName);
    } catch {
      if (!OtelMetricsTelemetrySink.importWarningEmitted) {
        OtelMetricsTelemetrySink.importWarningEmitted = true;
        console.warn(
          "@opentelemetry/api is not installed. OtelMetricsTelemetrySink is a no-op until the package is available.",
        );
      }
      return;
    }

    for (const decision of DECISION_WIRE_STRINGS) {
      this.decisionCounters.set(decision, meter.createCounter(`acs_intervention_${decision}_total`));
    }
    this.durationHistogram = meter.createHistogram("acs_intervention_duration_ms");
    this.isAvailable = true;
  }

  get available(): boolean {
    return this.isAvailable;
  }

  emit(event: TelemetryEvent): void {
    if (!this.isAvailable) {
      return;
    }
    // Record one increment and one duration sample per evaluation. Only the
    // base decision event records metrics, matching the Rust OTel sink, so a
    // non-decision event fed in directly cannot double-count.
    if (event.eventType !== TelemetryEventType.Decision) {
      return;
    }
    const attributes = otelAttributes(event);
    if (event.decision !== null) {
      this.decisionCounters.get(event.decision)?.add(1, attributes);
    }
    if (event.durationMs !== null) {
      this.durationHistogram?.record(event.durationMs, attributes);
    }
  }

  forceFlush(): void {
    if (!this.isAvailable) {
      return;
    }
    this.resolvedMeterProvider?.forceFlush?.();
  }

  shutdown(): void {
    if (!this.isAvailable) {
      return;
    }
    this.resolvedMeterProvider?.shutdown?.();
  }
}

export function coerceTelemetrySink(
  sink: TelemetrySink | readonly TelemetrySink[] | null | undefined,
): TelemetrySink | undefined {
  if (sink === null || sink === undefined) {
    return undefined;
  }
  if (Array.isArray(sink)) {
    return new MultiSink(sink.map((child) => requireTelemetrySink(child)));
  }
  return requireTelemetrySink(sink);
}

export function labelsFromClient(runtimeClient: unknown): {
  policyIds: Record<string, string>;
  annotators: Record<string, string[]>;
} {
  // Read the runtime client's policyLabels map, which the native client sources
  // from the merged manifest, so labels are populated on every native-backed
  // constructor including fromUrl and fromManifestChain and for extends-inherited
  // bindings. A client without policyLabels yields empty indexes, so policyId is
  // null and annotators fall back to executed annotation keys on the result.
  // Never throws, since telemetry labels are best effort.
  const policyIds: Record<string, string> = {};
  const annotators: Record<string, string[]> = {};
  let labels: unknown;
  try {
    const getter = (runtimeClient as { policyLabels?: unknown } | null)?.policyLabels;
    if (typeof getter !== "function") {
      return { policyIds, annotators };
    }
    labels = (getter as () => unknown).call(runtimeClient);
  } catch {
    return { policyIds, annotators };
  }
  if (!isRecord(labels)) {
    return { policyIds, annotators };
  }
  for (const [point, entry] of Object.entries(labels)) {
    if (!isRecord(entry)) {
      continue;
    }
    const pointKey = normalizeInterventionPointKey(point);
    if (typeof entry.policy_id === "string") {
      policyIds[pointKey] = entry.policy_id;
    }
    const names = entry.annotators;
    if (Array.isArray(names) && names.length > 0) {
      annotators[pointKey] = names.map((name) => String(name)).sort(compareUnicodeScalarKeys);
    }
  }
  return { policyIds, annotators };
}

function requireTelemetrySink(sink: unknown): TelemetrySink {
  if (!isObject(sink) || typeof (sink as { emit?: unknown }).emit !== "function") {
    throw new TypeError(
      `telemetrySink must be a TelemetrySink with an emit() method or an array of sinks, got ${typeName(sink)}`,
    );
  }
  return sink as TelemetrySink;
}

function otelAttributes(event: TelemetryEvent): Record<string, string> {
  const attributes: Record<string, string> = {
    event_type: event.eventType,
    intervention_point: String(event.interventionPoint),
  };
  if (event.enforcementMode !== null) attributes.enforcement_mode = event.enforcementMode;
  if (event.decision !== null) attributes.decision = event.decision;
  if (event.reasonCode !== null) attributes.reason_code = event.reasonCode;
  if (event.errorClass !== null) attributes.error_class = event.errorClass;
  if (event.policyId !== null) attributes.policy_id = event.policyId;
  if (event.annotators.length > 0) attributes.annotators = event.annotators.join(",");
  if (event.evidenceArtefact !== null) attributes.evidence_artefact = event.evidenceArtefact;
  if (event.evidenceVerificationPointerKeys.length > 0) {
    attributes.evidence_verification_pointer_keys = event.evidenceVerificationPointerKeys.join(",");
  }
  return attributes;
}

function annotatorNames(policyInput: JsonValue | undefined): string[] {
  if (!isRecord(policyInput) || !isRecord(policyInput.annotations)) {
    return [];
  }
  return Object.keys(policyInput.annotations).sort(compareUnicodeScalarKeys);
}

const INTERVENTION_POINT_NAME_TO_WIRE: Record<string, string> = {
  AgentStartup: "agent_startup",
  Input: "input",
  PreModelCall: "pre_model_call",
  PostModelCall: "post_model_call",
  PreToolCall: "pre_tool_call",
  PostToolCall: "post_tool_call",
  Output: "output",
  AgentShutdown: "agent_shutdown",
};

function normalizeInterventionPointKey(point: string): string {
  return INTERVENTION_POINT_NAME_TO_WIRE[point] ?? point;
}

function compareUnicodeScalarKeys(left: string, right: string): number {
  const leftScalars = Array.from(left);
  const rightScalars = Array.from(right);
  const length = Math.min(leftScalars.length, rightScalars.length);
  for (let index = 0; index < length; index += 1) {
    const leftCodePoint = leftScalars[index].codePointAt(0) ?? 0;
    const rightCodePoint = rightScalars[index].codePointAt(0) ?? 0;
    if (leftCodePoint !== rightCodePoint) return leftCodePoint - rightCodePoint;
  }
  return leftScalars.length - rightScalars.length;
}

function isRecord(value: unknown): value is Record<string, JsonValue> {
  return isObject(value) && !Array.isArray(value);
}

function isObject(value: unknown): value is object {
  return typeof value === "object" && value !== null;
}

function typeName(value: unknown): string {
  if (value === null) return "null";
  if (value === undefined) return "undefined";
  return value.constructor?.name ?? typeof value;
}
